docs/source/user/stencil.rst

```diff
@@ -0,0 +1,205 @@
+an index to be out-of-bounds then that element belongs to the border
+of the output array.  In the above example, points -1 and +1 are
+accessed in each dimension and thus the output array has a border
+of size one in all dimensions.
+
+
+Stencil decorator options
+=========================
+
+While the stencil decorator may be augmented in the future to 
+provide additional mechanisms for border handling, at the moment
+the stencil decorator currently supports only one option.
+
+.. _stencil-neighborhood:
+
+``neighborhood``
+----------------
+
+Sometimes it may be inconvenient to write the stencil kernel
+exclusively with Integer literals.  For example, let us say we
+would like to compute the trailing 30-day moving average of a
+time series of data.  One could write 
+``(a[-29] + a[-28] + ... + a[-1] + a[0]) / 30`` but the stencil
+decorator offers a more concise form using the ``neighborhood``
+option::
+
+   @stencil(neighborhood=((-29,0),))
+   def kernel2(a):
+       cumul = 0
+       for i in range(-29,1):
+           cumul += a[i]
+       return cumul / 30
+
+The neighborhood option is a tuple of tuples.  The outer tuple's
+length is equal to the number of dimensions of the input array.
+The inner tuples' lengths are always 2 because
+each element of the outer tuple corresponds to minimum and
+maximum index offsets used in the corresponding dimension.
+
+If a user specifies a neighborhood but then in the kernel 
+accesses elements outside the specified neighborhood, the behavior
+is undefined.
+
+.. _stencil-mode:
+
+``mode``
+--------
+
+The optional mode parameter controls how the border of the output array
+is handled.  Currently, there is only one supported value, ``"constant"``.
+In ``constant`` mode, the stencil kernel is not applied in cases where
+the kernel would access elements outside the valid range of the input
+array.  In such cases, those elements in the output array are assigned
+to a constant value, as specified by the ``cval`` parameter.
+
+``cval``
+--------
+
+The optional cval parameter defaults to zero but can be set to any
+desired value, which is then used for the border of the output array
+if the mode parameter is set to ``constant``.  The cval parameter is 
+ignored in all other modes.
+
+Stencil invocation options
+==========================
+
+Internally, the stencil decorator transforms the specified stencil
+kernel into a regular Python function.  In this process, the decorator
+adds one optional argument in addition to the first required parameter, 
+which is the input array.
+
+.. _stencil-function-out:
+
+``out``
+-------
+
+The optional ``out`` parameter is added to every stencil function
+generated by Numba.  If specified, the ``out`` parameter tells 
+Numba that the user is providing their own pre-allocated array 
+to be used for the output of the stencil.  In this case, the
+stencil function will not allocate its own output array.
+An example usage is shown below::
```

Comment:
Certainly we can do that.  I think I still prefer this approach rather than one where we coerce the result of the kernel to the type of the output array before assignment.